crypto/cms/cms_env.c

```diff
@@ -1337,6 +1366,17 @@ int ossl_cms_AuthEnvelopedData_final(CMS_ContentInfo *cms, BIO *cmsbio)
  */
 int ossl_cms_pkey_get_ri_type(EVP_PKEY *pk)
 {
+    int ri_type;
+    EVP_PKEY_CTX *ctx = NULL;
+
+    /*
+     * First check the provider for RecipientInfo support since a key may support
+     * multiple types, e.g. an RSA key and provider may support RSA key transport
+     * and/or RSA-KEM.
+     */
+    if (EVP_PKEY_get_int_param(pk, OSSL_PKEY_PARAM_CMS_RI_TYPE, &ri_type))
```

Comment:
I added the OSSL_PKEY_PARAM_CMS_RI_TYPE as a quick way for the provider to indicate what type of RecipientInfo its key supports.  I couldn't think of a better way of doing this.  Thoughts?  if the provider doesn't support this parameter, below we try hard to determine what type of RecipientInfo is supported.

OSSL_PKEY_PARAM_CMS_RI_TYPE could be change to return a list of RecipientInfo types the key supports, but I don't know if there's any value in that, we'd then have to choose among those types.  Might as well just let the provider supply a single type.